extmod,unix: Show compiler info in REPL banner.

stinos opened the pull request:
When testing out things I normally work in a single codebase, shared across different compiler/platform combinations (e.g. unix+clang+gcc, windows+msvc, migw+clang+gcc) which makes it easy to lose track of which micropython executable is being started. This fixes that just like CPython does.

It's not strictly needed to expose mp_platform_info_python_compiler_obj via a header, a less intrusive alternative would be for `do_repl` to get the `platform` module directly, but I'm not sure if that pattern is used somewhere else? Would look like:
```
extern const mp_obj_module_t mp_module_uplatform;
mp_obj_t fun = mp_obj_dict_get(mp_module_uplatform.globals, MP_ROM_QSTR(MP_QSTR_python_compiler));
mp_obj_t info_str = mp_call_function_0(fun);
...
```

Result looks like:
> MicroPython v1.18 on 2022-02-17; win32 [MSC v.1930 64 bit] version

or

> MicroPython v1.18 on 2022-02-17; win32 [GCC 11.2.0] version

dpgeorge commented:
This approach looks fine to me, it's OK to expose that uplatform object.

dpgeorge commented:
Maybe it makes more sense to move the definition of `PLATFORM_COMPILER` to `moduplatform.h` so it can be used directly?  That would be more efficient and use less code (and still be able to reuse the same string because the compiler is smart enough for that).

stinos commented:
Considered that yes, but shouldn't we then just move the other macros from .c to .h already as well (`PLATFORM_ARCH` etc)?

dpgeorge commented:
I don't see why not, that would be useful.  They'd need to be prefixed with something unique though, like `MICROPY_PLATFORM_xxx`?

stinos committed:
extmod/moduplatform: Move platform PP definitions into a header.

These are more generally useful than just for the module so make
them globally available, prefixed consistently with MICROPY_PLATFORM_.

stinos committed:
unix: Show compiler info in REPL banner.

The unix port's main.c gets used by unix and windows ports, and with
a variety of compilers, so it's convenient to see which version is
actually being used immediately when starting micropython.
This is similar to what CPython does.

dpgeorge commented on extmod/moduplatform.h:
```diff
@@ -0,0 +1,105 @@
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2022 Damien P. George
```
Would be better to copy the copyright line from `moduplatform.c`, because the macros below were copied out of that file.

stinos replied:
Ok.

dpgeorge commented:
Thanks for updating, merged in 49934fcf8b0b2bd906a96e1ff57b724d8dd6d93e and 5f50f4a130e60c4bc5f4fbf58ac849dcaf764e72
